### PR TITLE
Add option to save optimizer and scheduler state during training, and to resume training from these states

### LINKED ISSUE #3444: [Bug]: optimizer state not saved
### Describe the bug

Thank you for developing and maintaining this invaluable module!

We would like to save the state of the optimizer at the end of each epoch.
The `save_optimizer_state` parameter of the `fine_tune` function seems to be designed for this purpose.
However, the state of the optimizer is not saved even if we set `save_optimizer_state=True`.

Thank you!

### To Reproduce

```python
%pip install scipy==1.10.1 datasets transformers torch==2.0 flair==0.13.1 

import torch
import flair
from flair.data import Corpus
from flair.datasets import TREC_6
from flair.embeddings import TransformerDocumentEmbeddings
from flair.models import TextClassifier
from flair.trainers import ModelTrainer

# 1. get the corpus
corpus: Corpus = TREC_6()

# 2. what label do we want to predict?
label_type = 'question_class'

# 3. create the label dictionary
label_dict = corpus.make_label_dictionary(label_type=label_type)

# 4. initialize transformer document embeddings (many models are available)
document_embeddings = TransformerDocumentEmbeddings('distilbert-base-uncased', fine_tune=True)

# 5. create the text classifier
classifier = TextClassifier(document_embeddings, label_dictionary=label_dict, label_type=label_type)

# 6. initialize trainer
trainer = ModelTrainer(classifier, corpus)

# 7. run training with fine-tuning
trainer.fine_tune('resources/taggers/question-classification-with-transformer',
                  learning_rate=5.0e-5,
                  mini_batch_size=4,
                  max_epochs=10,
                  save_optimizer_state=True,
                  save_model_each_k_epochs=1
)

checkpoint = torch.load('resources/taggers/question-classification-with-transformer/model_epoch_1.pt', map_location=flair.device)
```

### Expected behavior

When `save_optimizer_state` is `true`, the checkpoint contains the state_dict of the optimizer.

### Logs and Stack traces

_No response_

### Screenshots

_No response_

### Additional Context

_No response_

##### Flair
0.13.1
##### Pytorch
2.0.0+cu117

### PR DESCRIPTION
**Pull request by alanakbik**
This PR adds functionality to save and load optimizer and scheduler states. You can enable this by setting `save_optimizer_state=True` in any of the train methods of `ModelTrainer` (train(), fine_tune(), or train_custom()). When enabled, both the optimizer state and the scheduler state are saved in the model file.

If you load a model that contains these states and pass it to the ModelTrainer, it will automatically load them if you continue training the model. 

Changes:
- Modified `ModelTrainer._save_model` to save scheduler states from active plugins
- Added scheduler_state_dict handling in Model._get_state_dict and load methods
- Updated scheduler plugins (`LinearScheduler` and `AnnealOnPlateau`) to load their states when resuming training

Closes #3444

### CODE REVIEW

**Comment by MattGPT-ai** [comment]
This is great! We are going to try using this when loading model checkpoints
